### PR TITLE
Keoken support for chain reorgs

### PR DESCRIPTION
**Pull request by fpelliccioni**

### CODE REVIEW

**Review by dario-ramos — DISMISSED** [review]

File: src/bitprim/keoken/memory_state.cpp [hunk 1]
```diff
@@ -29,15 +29,82 @@ using libbitcoin::wallet::payment_address;
...
+    size_t height;
+
+    explicit
+    rollback_pred(size_t height)
+        : height(height)
+    {}
+
+    template <typename Entry>
+    bool operator()(Entry const& entry) const {
+        return entry.block_height > height;
```
Comment (mdallago) [inline src/bitprim/keoken/memory_state.cpp]: `> or >=`

File: src/bitprim/keoken/memory_state.cpp [hunk 2]
```diff
@@ -29,15 +29,82 @@ using libbitcoin::wallet::payment_address;
...
+void memory_state::rollback_to(size_t height) {
+    boost::unique_lock<boost::shared_mutex> lock(mutex_);
+
+    remove_balance_entries(rollback_pred{height});
+
+    asset_list_.erase(
+        std::remove_if(begin(asset_list_), end(asset_list_), rollback_pred{height}), 
+    end(asset_list_));
+
+    asset_id_next_ = asset_list_.empty() ? asset_id_initial_ : asset_list_.back().asset.id();
```
Comment (mdallago) [inline src/bitprim/keoken/memory_state.cpp]: Missing asset_list_.back().asset.id() + 1 ?

Comment (fpelliccioni) [reply src/bitprim/keoken/memory_state.cpp]: Yeah, you are right! 
Good catch! 
Thanks!
`asset_id_next_ = asset_list_.empty() ? asset_id_initial_ : asset_list_.back().asset.id() + 1;`

File: src/bitprim/keoken/memory_state.cpp (hunk 1 above)
Comment (fpelliccioni) [reply src/bitprim/keoken/memory_state.cpp]: Really I don't know... I have to test the values...

**Review by mdallago — APPROVED** [review]

**Pull request merged**